{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "baba87e9-f9f0-4224-a583-cce87649d8ce",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Download of the dataset\n",
    "\n",
    "This notebook is used to to download the BacDive dataset. It is also available a .py version, useful to call the `download()` function externally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b376073-04fa-401f-94aa-792fafda228b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas\n",
    "import time\n",
    "import fnmatch\n",
    "import gzip\n",
    "import shutil\n",
    "import ftplib\n",
    "import numpy as np\n",
    "\n",
    "from ftplib import FTP, error_perm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84079108-8482-400e-a64e-2745b1ee2452",
   "metadata": {},
   "source": [
    "There are four mainly functions:\n",
    " * `prepare`: from some xlsx (specified in ifolder) it can retrieve the names of the files to download.\n",
    " * `download_ncbi_from_ftp`: given the genomes, the url, the suffix of the file (fna or gff) and a path, it downloads the files.\n",
    " * `download_patric_from_ftp`: download other files in a slightly different way.\n",
    " * `unpack_ncbi`: unpack the downloaded files using gzip.\n",
    " * `prepare_aflp_silco`: run the AFLP software and generate the .wri files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62730d59-2e3a-48fa-ae4d-19dde6bb6d32",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27d50c12-760e-4af5-a9d9-fb7141f06d0b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def prepare():\n",
    "    ifolder = '../Data/configuration/'\n",
    "    files_name = sorted([os.path.basename(f).replace('.xlsx', '') for f in glob.glob(ifolder + \"*.xlsx\")])\n",
    "\n",
    "    print(\"Preparing the download...\")\n",
    "    print(\"Custom download files:\", f\"\\n{files_name}\")\n",
    "    \n",
    "    if not os.path.exists('../Data/BacDive/'):\n",
    "        os.makedirs('../Data/BacDive/')    \n",
    "    \n",
    "    frames = []\n",
    "\n",
    "    for file_name in files_name:\n",
    "\n",
    "        file = pandas.read_excel(open(ifolder+file_name+'.xlsx', 'rb'),header=None) \n",
    "\n",
    "        file.columns = file.loc[9,:]\n",
    "        file.dropna(axis=0, how='all', inplace=True)\n",
    "        file.drop(file[(file['strains.Culture collection no.'].str.contains('http', na=False)) | \n",
    "                       (file['strains.Culture collection no.'] == 'URL')].index, inplace = True)\n",
    "        file.drop(file[file['strains.ID_strains'] == 'References'].index, inplace = True)\n",
    "        file.drop(file.head(5).index,inplace=True)\n",
    "        file.drop(file.tail(1).index,inplace=True)\n",
    "        file.drop(index=file.index[0], axis=0, inplace=True)\n",
    "\n",
    "        file.to_excel('../Data/BacDive/'+file_name+'.xlsx',index=None)\n",
    "\n",
    "        file_50CH = file.loc[file['kit_api_50CHac.ID_reference'] > 0]\n",
    "        file_50CH = file_50CH.iloc[:, np.r_[0:12,152:203]]\n",
    "        frames.append(file_50CH)\n",
    "\n",
    "    result_file_50CH = pandas.concat(frames)\n",
    "    result_file_50CH.to_excel('../Data/BacDive/Genome_50CH_ac.xlsx',index=None)\n",
    "\n",
    "    result_file_50CH = result_file_50CH.replace(['-', '+', '+/-'], [0,1,0])\n",
    "    result_file_50CH.to_excel('../Data/BacDive/Genome_50CH_ac_0_1.xlsx',index=None)\n",
    "    \n",
    "    print(\"** Generated files Genome_50CH_ac.xlsx and Genome_50CH_ac_0_1.xlsx into Data/BacDive/ folder. \\n\")\n",
    "    \n",
    "    genome_ncbi = result_file_50CH.loc[result_file_50CH['sequence_genomes.Sequence database'] == 'ncbi']\n",
    "    sequence_genomes_ncbi = list(set(genome_ncbi['sequence_genomes.Seq. accession number']))\n",
    "    \n",
    "    genome_patric = result_file_50CH.loc[result_file_50CH['sequence_genomes.Sequence database'] == 'patric']\n",
    "    sequence_genomes_patric = list(set(genome_patric['sequence_genomes.Seq. accession number']))\n",
    "    \n",
    "    return genome_ncbi, sequence_genomes_ncbi, genome_patric, sequence_genomes_patric \n",
    "\n",
    "\n",
    "\n",
    "def unpack_ncbi(folder_ncbi, files_name_ncbi, file_type):\n",
    "    print(f\"\\nUnzip files in {folder_ncbi}...\")\n",
    "    decompress_folder = os.path.join(folder_ncbi, 'Decompress')\n",
    "    \n",
    "    if not os.path.exists(decompress_folder):\n",
    "        os.makedirs(decompress_folder)\n",
    "\n",
    "    for file in files_name_ncbi:\n",
    "        input_file_path = os.path.join(folder_ncbi, file + file_type + '.gz')\n",
    "        output_file_path = os.path.join(decompress_folder, file + file_type)\n",
    "\n",
    "        with gzip.open(input_file_path, 'rb') as f_in:\n",
    "            with open(output_file_path, 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "                \n",
    "    print(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7bb2bb23-11d4-4747-a108-7f730abad278",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def download_ncbi_from_ftp(sequence_genomes, url_prefix, file_suffix, download_path):\n",
    "    tic = time.perf_counter()\n",
    "\n",
    "    ftp = ftplib.FTP('ftp.ncbi.nlm.nih.gov')\n",
    "    ftp.login()\n",
    "    \n",
    "    if not os.path.exists(download_path):\n",
    "        os.makedirs(download_path)\n",
    "        \n",
    "    for sequence in sequence_genomes:\n",
    "                \n",
    "        toc = time.perf_counter()\n",
    "\n",
    "        if toc - tic >= 200:\n",
    "            ftp.quit()\n",
    "            ftp = ftplib.FTP('ftp.ncbi.nlm.nih.gov')\n",
    "            ftp.login()\n",
    "\n",
    "            tic = time.perf_counter()\n",
    "\n",
    "        url = url_prefix\n",
    "\n",
    "        if sequence[0:3] == 'GCA':\n",
    "            for index in range(4, len(sequence), 3):\n",
    "                url = url + sequence[index:(index + 3)] + '/'\n",
    "\n",
    "        ftp.cwd(url)\n",
    "        name = next(iter(ftp.nlst()))\n",
    "        ftp.cwd(name)\n",
    "\n",
    "        files = ftp.nlst()\n",
    "                        \n",
    "        if not any(file_suffix in string for string in files):\n",
    "            print(f'  File {sequence} not found in GCA. Trying to find the .gff in GCF folder.')\n",
    "            url = url[:12] + 'GCF' + url[15:]\n",
    "            \n",
    "            ftp.cwd('/.')\n",
    "            try:\n",
    "                ftp.cwd(url)\n",
    "            except error_perm as e:\n",
    "                if \"550\" in str(e):\n",
    "                    print(\"\")\n",
    "                    print(\"=\"*80)\n",
    "                    print(f\"{url}: No such file or directory\")\n",
    "                    print(\"=\"*80,\"\\n\")     \n",
    "                else:\n",
    "                    return \" other error.\"\n",
    "                    \n",
    "            name = next(iter(ftp.nlst()))\n",
    "            ftp.cwd(name)\n",
    "\n",
    "            files = ftp.nlst()         \n",
    "            \n",
    "        # Download files\n",
    "        for file in files:\n",
    "\n",
    "            if fnmatch.fnmatch(file, name + file_suffix):  # To download specific files.\n",
    "\n",
    "                if file[0:3] == 'GCF': # Change file name from GCF_* to GCA_*.\n",
    "                    file_path = os.path.join(download_path, 'GCA' + file[3:])\n",
    "                else:\n",
    "                    file_path = os.path.join(download_path, file)\n",
    "                \n",
    "\n",
    "                if not os.path.exists(file_path):\n",
    "\n",
    "                    try:\n",
    "                        ftp.retrbinary(\"RETR \" + file, open(file_path, 'wb').write)\n",
    "\n",
    "                    except EOFError:  # To avoid EOF errors.\n",
    "                        pass\n",
    "                else:\n",
    "                    print(f\"  File already exists: {file_path}\")            \n",
    "                \n",
    "\n",
    "        ftp.cwd('/.')\n",
    "\n",
    "    ftp.quit()\n",
    "    \n",
    "    print(f\"Files downloaded in {download_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "144f4c61-942a-405f-95f1-fab03bb7125a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def download_patric_from_ftp(sequence_genomes, url_prefix, file_suffix, download_path):\n",
    "    tic = time.perf_counter()\n",
    "\n",
    "    ftp = ftplib.FTP('ftp.bvbrc.org')\n",
    "    ftp.login()\n",
    "    \n",
    "    if not os.path.exists(download_path):\n",
    "        os.makedirs(download_path)\n",
    "\n",
    "    # Unable to download entire sequence: \"no file or directory\" error when searching for data online.\n",
    "    sequence_error = [525325.5, 1589.1, 1589.9, 1226298.1, 'GCA_001475175', 219334.1, 391904.2, 1358.133, 51663.5, 1423723.1, 273384.3, 518634.2]\n",
    "    \n",
    "    for sequence in sequence_genomes:\n",
    "\n",
    "        toc = time.perf_counter()\n",
    "\n",
    "        if toc - tic >= 200:\n",
    "            ftp.quit()\n",
    "            ftp = ftplib.FTP('ftp.bvbrc.org')\n",
    "            ftp.login()\n",
    "\n",
    "            tic = time.perf_counter()\n",
    "\n",
    "        if sequence in sequence_error:\n",
    "            print(f'  *** Strain {sequence} not downloaded: file not found online. ***\\n')\n",
    "\n",
    "        else:\n",
    "            url = url_prefix + str(sequence) + '/'\n",
    "            ftp.cwd(url)\n",
    "\n",
    "            file_name = str(sequence) + file_suffix\n",
    "\n",
    "            files = ftp.nlst()\n",
    "\n",
    "            # Download files\n",
    "            for file in files:\n",
    "                if fnmatch.fnmatch(file, file_name):  # To download specific files.\n",
    "                    file_path = os.path.join(download_path, file)\n",
    "                    \n",
    "                    if not os.path.exists(file_path):\n",
    "\n",
    "                        try:\n",
    "                            ftp.retrbinary(\"RETR \" + file, open(file_path, 'wb').write)\n",
    "\n",
    "                        except EOFError:  # To avoid EOF errors.\n",
    "                            pass\n",
    "                    else:\n",
    "                        print(f\"  File already exists: {file_path}\")\n",
    "\n",
    "            ftp.cwd('/.')\n",
    "\n",
    "    ftp.quit()\n",
    "  \n",
    "    print(f\"Files downloaded in {download_path}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e3b3916-7c35-4c9e-a90b-1e11eb3341ad",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Download files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "47f09917-cbe7-44e6-8db6-7ad6fcff3d80",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def download(genome_ncbi, sequence_genomes_ncbi, genome_patric, sequence_genomes_patric):\n",
    "    url_prefix = 'genomes/all/GCA/'\n",
    "    file_suffix_fna = '_genomic.fna.gz'\n",
    "    download_path_fna = '../Data/BacDive/ncbi/'\n",
    "    file_type_fna = '.fna'\n",
    "\n",
    "    file_suffix_gff = '_genomic.gff.gz'\n",
    "    download_path_gff = '../Data/BacDive/gff/ncbi/'\n",
    "    file_type_gff = '.gff'\n",
    "\n",
    "    patric_url_prefix_fna = 'genomes/'\n",
    "    patric_file_suffix_fna = '.fna'\n",
    "    patric_download_path_fna = '../Data/BacDive/Patric/'\n",
    "\n",
    "    patric_url_prefix_gff = 'genomes/'\n",
    "    patric_file_suffix_gff = '.PATRIC.gff'\n",
    "    patric_download_path_gff = '../Data/BacDive/gff/Patric/'\n",
    "\n",
    "    paths = pandas.Series([download_path_fna, download_path_gff, patric_download_path_fna, patric_download_path_gff])\n",
    "\n",
    "    for path in paths:\n",
    "        if not os.path.exists(path):\n",
    "            os.makedirs(path)\n",
    "\n",
    "\n",
    "    try:\n",
    "        print(\"\\nDownloading ncbi fna files...\")\n",
    "        download_ncbi_from_ftp(sequence_genomes_ncbi, url_prefix, file_suffix_fna, download_path_fna)\n",
    "\n",
    "        print(\"\\nDownloading ncbi gff files... (Problemi)\")\n",
    "        download_ncbi_from_ftp(sequence_genomes_ncbi, url_prefix, file_suffix_gff, download_path_gff)\n",
    "        \n",
    "        print(\"\\nDownloading Patric fna files...\")\n",
    "        download_patric_from_ftp(sequence_genomes_patric, patric_url_prefix_fna, patric_file_suffix_fna, patric_download_path_fna)\n",
    "        \n",
    "        print(\"\\nDownloading Patric gff files...\")\n",
    "        download_patric_from_ftp(sequence_genomes_patric, patric_url_prefix_gff, patric_file_suffix_gff, patric_download_path_gff)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(\"=\"*80)\n",
    "        print(f\"Errore: {e}\")\n",
    "        print(\"=\"*80)\n",
    "        print('\\n')\n",
    "        \n",
    "        \n",
    "    files_name_ncbi_fna = sorted([f.replace('.fna.gz', '').replace(download_path_fna, '').replace(download_path_fna, '') for f in glob.glob(download_path_fna + \"*.fna.gz\")])\n",
    "    files_name_ncbi_gff = sorted([f.replace('.gff.gz', '').replace(download_path_gff, '').replace(download_path_gff, '') for f in glob.glob(download_path_gff + \"*.gff.gz\")])\n",
    "\n",
    "    unpack_ncbi(download_path_fna, files_name_ncbi_fna, file_type_fna)\n",
    "    unpack_ncbi(download_path_gff, files_name_ncbi_gff, file_type_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "422cfccf-63d9-4274-8460-c3307bb5a0b7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Downloading ncbi fna files...\n",
      "\n",
      "Downloading ncbi gff files... (Problemi)\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007992275.1_ASM799227v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001025155.1_ASM102515v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_019753765.1_ASM1975376v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009764355.1_ASM976435v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_018314255.1_ASM1831425v1_genomic.gff.gz\n",
      "  File GCA_010120595 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_010120595.1_ASM1012059v1_genomic.gff.gz\n",
      "  File GCA_000374385 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000374385.1_ASM37438v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001025175.1_ASM102517v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017874575.1_ASM1787457v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014898155.1_ASM1489815v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_019586025.1_ASM1958602v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014207505.1_ASM1420750v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002706705.1_ASM270670v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009764365.1_L._kimchii_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_023657385.1_ASM2365738v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900079995.1_ASM90007999v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003697165.1_ASM369716v1_genomic.gff.gz\n",
      "  File GCA_902833925 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_902833925.1_ASM90283392v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000024285.1_ASM2428v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000829035.1_ASM82903v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001050115.1_ASM105011v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003641145.1_ASM364114v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002095475.1_ASM209547v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900015005.1_Lactobacillus_oligofermentans_LMG_22743_genome_genomic.gff.gz\n",
      "  File GCA_001311115 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001311115.1_ASM131111v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000469365.1_ASM46936v1_genomic.gff.gz\n",
      "  File GCA_000427785 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000427785.1_ASM42778v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009727155.1_ASM972715v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001042695.1_ASM104269v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001274725.1_ASM127472v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002706375.1_ASM270637v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009299505.1_ASM929950v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000829055.1_ASM82905v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_018717645.1_ASM1871764v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001888925.1_ASM188892v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014648715.1_ASM1464871v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007991135.1_ASM799113v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000196555.1_ASM19655v1_genomic.gff.gz\n",
      "  File GCA_000744825 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000744825.1_ASM74482v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002907375.1_ASM290737v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001025195.1_ASM102519v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007992235.1_ASM799223v1_genomic.gff.gz\n",
      "  File GCA_019655995 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_019655995.1_ASM1965599v1_genomic.gff.gz\n",
      "  File GCA_001591685 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001591685.1_ASM159168v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000829395.1_ASM82939v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900637145.1_49569_E01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001047135.1_ASM104713v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_008326545.1_ASM832654v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001888965.1_ASM188896v1_genomic.gff.gz\n",
      "  File GCA_000425865 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000425865.1_ASM42586v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000160855.1_ASM16085v1_genomic.gff.gz\n",
      "  File GCA_000615725 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000615725.1_ASM61572v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000163475.1_ASM16347v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014635685.1_ASM1463568v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_023657365.1_ASM2365736v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900637335.1_50618_H02_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900089835.1_VthalassaeCECT8203_Velvet_Prokka_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900103055.1_IMG-taxon_2636416037_annotated_assembly_genomic.gff.gz\n",
      "  File GCA_000615245 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000615245.1_ASM61524v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000014425.1_ASM1442v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002370355.1_ASM237035v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001047695.1_ASM104769v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007991875.1_ASM799187v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002259645.1_ASM225964v1_genomic.gff.gz\n",
      "  File GCA_905371965 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "================================================================================\n",
      "genomes/all/GCF/905/371/965/: No such file or directory\n",
      "================================================================================\n",
      "\n",
      "  File GCA_019656055 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_019656015 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_001591885 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003795105.1_ASM379510v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_018094625.1_ASM1809462v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009299485.1_ASM929948v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001433985.1_ASM143398v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900637175.1_49569_F01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002811195.1_ASM281119v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014646835.1_ASM1464683v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002217925.1_ASM221792v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002849895.1_ASM284989v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014651215.1_ASM1465121v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017132775.1_ASM1713277v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001025215.1_ASM102521v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900187315.1_52087_C01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014651255.1_ASM1465125v1_genomic.gff.gz\n",
      "  File GCA_003946245 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014467235.1_ASM1446723v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_016908715.1_ASM1690871v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900167245.1_IMG-taxon_2568526012_annotated_assembly_genomic.gff.gz\n",
      "  File GCA_003946275 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003946275.1_ASM394627v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_020089505.1_ASM2008950v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_012103035.1_O.oryzae_NBRC_102588.fasta_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003641185.1_ASM364118v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000296855.1_ASM29685v1_genomic.gff.gz\n",
      "  File GCA_019656075 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017654245.1_ASM1765424v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_018987235.1_ASM1898723v1_genomic.gff.gz\n",
      "  File GCA_001311505 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001436025.1_ASM143602v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000706765.1_ASM70676v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003423665.1_ASM342366v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000876205.1_ASM87620v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000056065.1_ASM5606v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000498955.1_DSM-21115_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_019444005.1_ASM1944400v1_genomic.gff.gz\n",
      "  File GCA_001313225 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000632805.1_ASM63280v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000196855.1_ASM19685v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007990205.1_ASM799020v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001437255.1_ASM143725v1_genomic.gff.gz\n",
      "  File GCA_000185065 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000831645.1_ASM83164v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001465545.1_ASM146554v1_genomic.gff.gz\n",
      "  File GCA_000615325 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000615325.1_ASM61532v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001908495.1_ASM190849v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_004209755.1_ASM420975v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_013394085.1_ASM1339408v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000737845.1_BBOMBassembly_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014191315.1_ASM1419131v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001042615.1_ASM104261v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_004322755.1_ASM432275v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900476045.1_50279_D01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900637095.1_48450_G02_genomic.gff.gz\n",
      "  File GCA_900324475 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900324475.1_EB-247_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000269965.1_ASM26996v1_genomic.gff.gz\n",
      "  File GCA_000426405 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000426405.1_ASM42640v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007990855.1_ASM799085v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003096575.1_ASM309657v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014651415.1_ASM1465141v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017298735.1_ASM1729873v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007989165.1_ASM798916v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003047065.1_ASM304706v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002706745.1_ASM270674v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000010425.1_ASM1042v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007989145.1_ASM798914v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900187025.1_49923_F01_genomic.gff.gz\n",
      "  File GCA_000744255 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002015125.1_ASM201512v1_genomic.gff.gz\n",
      "  File GCA_019655915 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001263395.1_ASM126339v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002706665.1_ASM270666v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000967195.1_ASM96719v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002736025.1_ASM273602v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002259755.1_ASM225975v1_genomic.gff.gz\n",
      "  File GCA_014634745 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_003946015 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003946015.1_ASM394601v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002259705.1_ASM225970v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002706425.1_ASM270642v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002259605.1_ASM225960v1_genomic.gff.gz\n",
      "  File GCA_000615405 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_000429585 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900186835.1_49950_F01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_901212485.1_NZ_Bifidobacterium_bifidum_MC1_genomic.gff.gz\n",
      "  File GCA_000771045 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_002916915 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002916915.1_ASM291691v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000014445.1_ASM1444v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000008565.1_ASM856v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007989125.1_ASM798912v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900478165.1_48290_A02_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009299475.1_ASM929947v1_genomic.gff.gz\n",
      "  File GCA_000423725 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001437285.1_ASM143728v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_004362535.1_ASM436253v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001025135.1_ASM102513v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000020425.1_ASM2042v1_genomic.gff.gz\n",
      "  File GCA_000166715 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002259745.1_ASM225974v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014651235.1_ASM1465123v1_genomic.gff.gz\n",
      "  File GCA_001039045 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001543205.1_ASM154320v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001908415.1_ASM190841v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003641165.1_ASM364116v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009495475.1_ASM949547v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002278095.1_ASM227809v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000010005.1_ASM1000v1_genomic.gff.gz\n",
      "  File GCA_905371805 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_905371805.1_SRR9217392-mag-bin.4_genomic.gff.gz\n",
      "  File GCA_005405505 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000241055.1_ASM24105v1_genomic.gff.gz\n",
      "  File GCA_000615205 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014898175.1_ASM1489817v1_genomic.gff.gz\n",
      "  File GCA_001311335 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001311335.1_ASM131133v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_004101845.1_ASM410184v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001042675.1_ASM104267v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900637215.1_49888_B01_genomic.gff.gz\n",
      "  File GCA_001311525 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001311525.1_ASM131152v1_genomic.gff.gz\n",
      "  File GCA_002607735 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000260715.1_ASM26071v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_003369815.1_ASM336981v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900103655.1_IMG-taxon_2597490363_annotated_assembly_genomic.gff.gz\n",
      "  File GCA_001592085 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_012396505.1_ASM1239650v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014131735.1_ASM1413173v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007991995.1_ASM799199v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_009764345.1_ASM976434v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007991295.1_ASM799129v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007989185.1_ASM798918v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001436135.1_ASM143613v1_genomic.gff.gz\n",
      "  File GCA_000771085 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007991855.1_ASM799185v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001434175.1_ASM143417v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900187015.1_50465_F01_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017299435.1_ASM1729943v1_genomic.gff.gz\n",
      "  File GCA_019655935 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_005405645 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000146325.1_ASM14632v1_genomic.gff.gz\n",
      "  File GCA_005405385 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File GCA_002777965 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_015698225.1_ASM1569822v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014651435.1_ASM1465143v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_007992315.1_ASM799231v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001042635.1_ASM104263v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_016908275.1_ASM1690827v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_014635905.1_ASM1463590v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001281175.1_ASM128117v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002907365.1_ASM290736v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000016825.1_ASM1682v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001936335.1_ASM193633v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001891125.1_ASM189112v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_900099625.1_IMG-taxon_2597490350_annotated_assembly_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017132755.1_ASM1713275v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000143745.1_ASM14374v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_002802875.1_Spades_TRE1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_019443985.1_ASM1944398v1_genomic.gff.gz\n",
      "  File GCA_000745125 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000745125.1_ASM74512v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000970855.1_ASM97085v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_017086525.1_ASM1708652v1_genomic.gff.gz\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_001436735.1_ASM143673v1_genomic.gff.gz\n",
      "  File GCA_000423745 not found in GCA. Trying to find the .gff in GCF folder.\n",
      "  File already exists: ../Data/BacDive/gff/ncbi/GCA_000423745.1_ASM42374v1_genomic.gff.gz\n",
      "Files downloaded in ../Data/BacDive/gff/ncbi/\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "-1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_ncbi, sequence_genomes_ncbi, genome_patric, sequence_genomes_patric = prepare() # Prepare all the files to download\n",
    "\n",
    "download(genome_ncbi, sequence_genomes_ncbi, genome_patric, sequence_genomes_patric)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "881fd9df-d894-4626-bfb5-7643982f82d4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Check missing files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "af04bc56-d660-4679-9d90-8f1d3c3acd01",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nomi di file presenti solo nella cartella ../Data/BacDive/ncbi/Decompress/:\n",
      "GCA_000185065.2_ASM18506v1_gen\n",
      "GCA_905371965.1_SRR9217392-mag-bin.15_gen\n",
      "GCA_000166715.2_ASM16671v1_gen\n",
      "\n",
      "\n",
      "\n",
      "Nomi di file presenti solo nella cartella ../Data/BacDive/gff/ncbi/Decompress/:\n",
      "GCA_000166715.1_ASM16671v1_gen\n",
      "GCA_000185065.1_ASM18506v1_gen\n"
     ]
    }
   ],
   "source": [
    "folder1 = '../Data/BacDive/ncbi/Decompress/'\n",
    "folder2 = '../Data/BacDive/gff/ncbi/Decompress/'\n",
    "\n",
    "files1 = [os.path.splitext(f)[0][:-4] for f in os.listdir(folder1) if os.path.isfile(os.path.join(folder1, f))]\n",
    "files2 = [os.path.splitext(f)[0][:-4] for f in os.listdir(folder2) if os.path.isfile(os.path.join(folder2, f))]\n",
    "\n",
    "unique_files_in_folder1 = set(files1) - set(files2)\n",
    "unique_files_in_folder2 = set(files2) - set(files1)\n",
    "\n",
    "print(f\"Nomi di file presenti solo nella cartella {folder1}:\")\n",
    "for filename in unique_files_in_folder1:\n",
    "    print(filename)\n",
    "    \n",
    "print('\\n\\n')\n",
    "print(f\"Nomi di file presenti solo nella cartella {folder2}:\")\n",
    "for filename in unique_files_in_folder2:\n",
    "    print(filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33c66883-07f1-45f2-aa59-4d6fd4a23927",
   "metadata": {},
   "source": [
    "# Generate the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54bc910-ce46-451c-b38f-ca545360acfc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_001025135.1_ASM102513v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_023657385.1_ASM2365738v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009764365.1_L._kimchii_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007989125.1_ASM798912v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002706425.1_ASM270642v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001543205.1_ASM154320v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000745125.1_ASM74512v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000241055.1_ASM24105v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000425865.1_ASM42586v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000498955.1_DSM-21115_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900476045.1_50279_D01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000196555.1_ASM19655v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019655995.1_ASM1965599v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002907375.1_ASM290737v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002259605.1_ASM225960v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000024285.1_ASM2428v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001039045.1_ASM103904v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000426405.1_ASM42640v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017086525.1_ASM1708652v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000831645.1_ASM83164v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000196855.1_ASM19685v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_905371965.1_SRR9217392-mag-bin.15_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900478165.1_48290_A02_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_004322755.1_ASM432275v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_023657365.1_ASM2365736v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001436135.1_ASM143613v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017654245.1_ASM1765424v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014191315.1_ASM1419131v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000706765.1_ASM70676v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000615205.1_ASM61520v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002777965.1_ASM277796v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001025195.1_ASM102519v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900637095.1_48450_G02_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007989185.1_ASM798918v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000970855.1_ASM97085v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003795105.1_ASM379510v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002736025.1_ASM273602v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_905371805.1_SRR9217392-mag-bin.4_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001591685.1_ASM159168v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000615725.1_ASM61572v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900103655.1_IMG-taxon_2597490363_annotated_assembly_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001311505.1_ASM131150v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900187315.1_52087_C01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002811195.1_ASM281119v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001891125.1_ASM189112v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000771085.1_DSM-19703_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002706745.1_ASM270674v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001311525.1_ASM131152v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019655915.1_ASM1965591v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007990855.1_ASM799085v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900187015.1_50465_F01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003946275.1_ASM394627v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_005405645.1_ASM540564v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003641185.1_ASM364118v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003423665.1_ASM342366v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002706665.1_ASM270666v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019655935.1_ASM1965593v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007991295.1_ASM799129v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_018717645.1_ASM1871764v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014646835.1_ASM1464683v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014467235.1_ASM1446723v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000020425.1_ASM2042v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001465545.1_ASM146554v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019753765.1_ASM1975376v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001047695.1_ASM104769v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900324475.1_EB-247_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002916915.1_ASM291691v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900637145.1_49569_E01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000737845.1_BBOMBassembly_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000615405.1_ASM61540v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_016908275.1_ASM1690827v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009764355.1_ASM976435v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001042675.1_ASM104267v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_016908715.1_ASM1690871v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000876205.1_ASM87620v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000160855.1_ASM16085v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_008326545.1_ASM832654v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001591885.1_ASM159188v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014635685.1_ASM1463568v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003946015.1_ASM394601v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_005405505.1_ASM540550v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001888925.1_ASM188892v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000829035.1_ASM82903v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007992275.1_ASM799227v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002259755.1_ASM225975v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_902833925.1_ASM90283392v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001936335.1_ASM193633v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900099625.1_IMG-taxon_2597490350_annotated_assembly_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000014445.1_ASM1444v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000010425.1_ASM1042v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000427785.1_ASM42778v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000014425.1_ASM1442v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900103055.1_IMG-taxon_2636416037_annotated_assembly_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001050115.1_ASM105011v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002802875.1_Spades_TRE1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019443985.1_ASM1944398v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900637215.1_49888_B01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002095475.1_ASM209547v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000829055.1_ASM82905v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007991995.1_ASM799199v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002259705.1_ASM225970v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002849895.1_ASM284989v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001592085.1_ASM159208v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007990205.1_ASM799020v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001311335.1_ASM131133v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_012396505.1_ASM1239650v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900187025.1_49923_F01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_004209755.1_ASM420975v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000469365.1_ASM46936v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_018094625.1_ASM1809462v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017299435.1_ASM1729943v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007989145.1_ASM798914v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900089835.1_VthalassaeCECT8203_Velvet_Prokka_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000423725.1_ASM42372v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002259645.1_ASM225964v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019656055.1_ASM1965605v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_012103035.1_O.oryzae_NBRC_102588.fasta_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001025155.1_ASM102515v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007992315.1_ASM799231v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001313225.1_ASM131322v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_010120595.1_ASM1012059v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017874575.1_ASM1787457v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009299505.1_ASM929950v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_018987235.1_ASM1898723v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001908415.1_ASM190841v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000423745.1_ASM42374v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000296855.1_ASM29685v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001437255.1_ASM143725v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002907365.1_ASM290736v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900167245.1_IMG-taxon_2568526012_annotated_assembly_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014651235.1_ASM1465123v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007989165.1_ASM798916v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_005405385.1_ASM540538v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001025175.1_ASM102517v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014898175.1_ASM1489817v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014651415.1_ASM1465141v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003946245.1_ASM394624v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001908495.1_ASM190849v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900079995.1_ASM90007999v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002706705.1_ASM270670v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002217925.1_ASM221792v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001433985.1_ASM143398v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000374385.1_ASM37438v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007991875.1_ASM799187v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000771045.1_DSM-17755_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017132775.1_ASM1713277v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014651255.1_ASM1465125v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000615245.1_ASM61524v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001434175.1_ASM143417v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000269965.1_ASM26996v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_018314255.1_ASM1831425v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001025215.1_ASM102521v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014207505.1_ASM1420750v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_901212485.1_NZ_Bifidobacterium_bifidum_MC1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000016825.1_ASM1682v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003096575.1_ASM309657v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007991855.1_ASM799185v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000829395.1_ASM82939v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000010005.1_ASM1000v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_015698225.1_ASM1569822v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_013394085.1_ASM1339408v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014635905.1_ASM1463590v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002706375.1_ASM270637v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001311115.1_ASM131111v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001437285.1_ASM143728v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003369815.1_ASM336981v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_004362535.1_ASM436253v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001281175.1_ASM128117v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001436025.1_ASM143602v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001042635.1_ASM104263v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000185065.2_ASM18506v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900186835.1_49950_F01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000429585.1_ASM42958v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019656015.1_ASM1965601v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017298735.1_ASM1729873v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001047135.1_ASM104713v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000632805.1_ASM63280v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000056065.1_ASM5606v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002278095.1_ASM227809v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009299485.1_ASM929948v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003697165.1_ASM369716v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900637175.1_49569_F01_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000166715.2_ASM16671v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001042615.1_ASM104261v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900637335.1_50618_H02_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007991135.1_ASM799113v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_017132755.1_ASM1713275v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000260715.1_ASM26071v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000744255.1_ASM74425v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_004101845.1_ASM410184v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003047065.1_ASM304706v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000143745.1_ASM14374v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009764345.1_ASM976434v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014651435.1_ASM1465143v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002607735.1_ASM260773v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014898155.1_ASM1489815v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002015125.1_ASM201512v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001042695.1_ASM104269v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000146325.1_ASM14632v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009727155.1_ASM972715v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002259745.1_ASM225974v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000615325.1_ASM61532v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014651215.1_ASM1465121v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019586025.1_ASM1958602v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009495475.1_ASM949547v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_020089505.1_ASM2008950v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001263395.1_ASM126339v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_900015005.1_Lactobacillus_oligofermentans_LMG_22743_genome_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003641145.1_ASM364114v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000967195.1_ASM96719v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014634745.1_ASM1463474v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001436735.1_ASM143673v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_007992235.1_ASM799223v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019444005.1_ASM1944400v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014648715.1_ASM1464871v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000163475.1_ASM16347v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_014131735.1_ASM1413173v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000008565.1_ASM856v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001274725.1_ASM127472v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_019656075.1_ASM1965607v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_003641165.1_ASM364116v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_001888965.1_ASM188896v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_009299475.1_ASM929947v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_000744825.1_ASM74482v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "GCA_002370355.1_ASM237035v1_genomic\n",
      "selectiv5 \tselectiv3 \n",
      "1423804.4\n",
      "selectiv5 \tselectiv3 \n",
      "348151.3\n",
      "selectiv5 \tselectiv3 \n",
      "1121112.3\n",
      "selectiv5 \tselectiv3 \n",
      "941770.3\n",
      "selectiv5 \tselectiv3 \n",
      "525327.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423748.3\n",
      "selectiv5 \tselectiv3 \n",
      "566552.18\n",
      "selectiv5 \tselectiv3 \n",
      "1236940.3\n",
      "selectiv5 \tselectiv3 \n",
      "1122147.4\n",
      "selectiv5 \tselectiv3 \n",
      "1217420.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423791.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423736.4\n",
      "selectiv5 \tselectiv3 \n",
      "1437600.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423797.16\n",
      "selectiv5 \tselectiv3 \n",
      "367928.6\n",
      "selectiv5 \tselectiv3 \n",
      "649758.3\n",
      "selectiv5 \tselectiv3 \n",
      "1046597.4\n",
      "selectiv5 \tselectiv3 \n",
      "1437606.5\n",
      "selectiv5 \tselectiv3 \n",
      "1423820.4\n",
      "selectiv5 \tselectiv3 \n",
      "375175.92\n",
      "selectiv5 \tselectiv3 \n",
      "1423797.7\n",
      "selectiv5 \tselectiv3 \n",
      "1236534.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423753.3\n",
      "selectiv5 \tselectiv3 \n",
      "561180.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423800.3\n",
      "selectiv5 \tselectiv3 \n",
      "1293592.9\n",
      "selectiv5 \tselectiv3 \n",
      "987066.3\n",
      "selectiv5 \tselectiv3 \n",
      "1121109.3\n",
      "selectiv5 \tselectiv3 \n",
      "1223498.3\n",
      "selectiv5 \tselectiv3 \n",
      "47770.64\n",
      "selectiv5 \tselectiv3 \n",
      "525325.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423768.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423777.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423819.3\n",
      "selectiv5 \tselectiv3 \n",
      "1203066.3\n",
      "selectiv5 \tselectiv3 \n",
      "872328.3\n",
      "selectiv5 \tselectiv3 \n",
      "525364.3\n",
      "selectiv5 \tselectiv3 \n",
      "649758.4\n",
      "selectiv5 \tselectiv3 \n",
      "525328.13\n",
      "selectiv5 \tselectiv3 \n",
      "1423736.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423773.3\n",
      "selectiv5 \tselectiv3 \n",
      "525338.16\n",
      "selectiv5 \tselectiv3 \n",
      "1423802.4\n",
      "selectiv5 \tselectiv3 \n",
      "525338.3\n",
      "selectiv5 \tselectiv3 \n",
      "1122147.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423757.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423723.12\n",
      "selectiv5 \tselectiv3 \n",
      "321983.3\n",
      "selectiv5 \tselectiv3 \n",
      "942150.3\n",
      "selectiv5 \tselectiv3 \n",
      "391904.5\n",
      "selectiv5 \tselectiv3 \n",
      "1423730.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423775.4\n",
      "selectiv5 \tselectiv3 \n",
      "941770.4\n",
      "selectiv5 \tselectiv3 \n",
      "1620.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423732.4\n",
      "selectiv5 \tselectiv3 \n",
      "1293598.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423768.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423806.3\n",
      "selectiv5 \tselectiv3 \n",
      "216816.128\n",
      "selectiv5 \tselectiv3 \n",
      "762211.3\n",
      "selectiv5 \tselectiv3 \n",
      "187452.24\n",
      "selectiv5 \tselectiv3 \n",
      "1236942.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423734.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423771.3\n",
      "selectiv5 \tselectiv3 \n",
      "566552.4\n",
      "selectiv5 \tselectiv3 \n",
      "1226675.3\n",
      "selectiv5 \tselectiv3 \n",
      "219334.4\n",
      "selectiv5 \tselectiv3 \n",
      "525366.3\n",
      "selectiv5 \tselectiv3 \n",
      "1395513.3\n",
      "selectiv5 \tselectiv3 \n",
      "1255.112\n",
      "selectiv5 \tselectiv3 \n",
      "1689.18\n",
      "selectiv5 \tselectiv3 \n",
      "1123358.5\n",
      "selectiv5 \tselectiv3 \n",
      "1114972.5\n",
      "selectiv5 \tselectiv3 \n",
      "1423792.3\n",
      "selectiv5 \tselectiv3 \n",
      "1291734.4\n",
      "selectiv5 \tselectiv3 \n",
      "1046596.6\n",
      "selectiv5 \tselectiv3 \n",
      "1423752.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423717.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423823.4\n",
      "selectiv5 \tselectiv3 \n",
      "1590.1803\n",
      "selectiv5 \tselectiv3 \n",
      "219334.11\n",
      "selectiv5 \tselectiv3 \n",
      "500634.6\n",
      "selectiv5 \tselectiv3 \n",
      "1423776.3\n",
      "selectiv5 \tselectiv3 \n",
      "1150460.3\n",
      "selectiv5 \tselectiv3 \n",
      "585520.12\n",
      "selectiv5 \tselectiv3 \n",
      "1588.22\n",
      "selectiv5 \tselectiv3 \n",
      "1423769.4\n",
      "selectiv5 \tselectiv3 \n",
      "881260.84\n",
      "selectiv5 \tselectiv3 \n",
      "585524.9\n",
      "selectiv5 \tselectiv3 \n",
      "1236943.3\n",
      "selectiv5 \tselectiv3 \n",
      "888027.3\n",
      "selectiv5 \tselectiv3 \n",
      "1587.663\n",
      "selectiv5 \tselectiv3 \n",
      "1114972.6\n",
      "selectiv5 \tselectiv3 \n",
      "1423749.3\n",
      "selectiv5 \tselectiv3 \n",
      "1236941.3\n",
      "selectiv5 \tselectiv3 \n",
      "930942.3\n",
      "selectiv5 \tselectiv3 \n",
      "1236939.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423789.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423774.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423731.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423805.4\n",
      "selectiv5 \tselectiv3 \n",
      "33965.25\n",
      "selectiv5 \tselectiv3 \n",
      "1423750.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423715.3\n",
      "selectiv5 \tselectiv3 \n",
      "500634.4\n",
      "selectiv5 \tselectiv3 \n",
      "1046596.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423821.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423790.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423728.3\n",
      "selectiv5 \tselectiv3 \n",
      "1133569.4\n",
      "selectiv5 \tselectiv3 \n",
      "396598.11\n",
      "selectiv5 \tselectiv3 \n",
      "616990.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423807.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423717.5\n",
      "selectiv5 \tselectiv3 \n",
      "1423794.3\n",
      "selectiv5 \tselectiv3 \n",
      "1358.88\n",
      "selectiv5 \tselectiv3 \n",
      "1423776.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423733.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423823.3\n",
      "selectiv5 \tselectiv3 \n",
      "1226298.3\n",
      "selectiv5 \tselectiv3 \n",
      "1579.22\n",
      "selectiv5 \tselectiv3 \n",
      "1291734.3\n",
      "selectiv5 \tselectiv3 \n",
      "1348632.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423801.4\n",
      "selectiv5 \tselectiv3 \n",
      "324831.46\n",
      "selectiv5 \tselectiv3 \n",
      "1437603.4\n",
      "selectiv5 \tselectiv3 \n",
      "907931.3\n",
      "selectiv5 \tselectiv3 \n",
      "1329411.4\n",
      "selectiv5 \tselectiv3 \n",
      "83683.14\n",
      "selectiv5 \tselectiv3 \n",
      "1590.625\n",
      "selectiv5 \tselectiv3 \n",
      "525365.3\n",
      "selectiv5 \tselectiv3 \n",
      "33966.4\n",
      "selectiv5 \tselectiv3 \n",
      "1150460.13\n",
      "selectiv5 \tselectiv3 \n",
      "500634.3\n",
      "selectiv5 \tselectiv3 \n",
      "1267156.3\n",
      "selectiv5 \tselectiv3 \n",
      "585520.17\n",
      "selectiv5 \tselectiv3 \n",
      "45462.11\n",
      "selectiv5 \tselectiv3 \n",
      "1437607.3\n",
      "selectiv5 \tselectiv3 \n",
      "888027.6\n",
      "selectiv5 \tselectiv3 \n",
      "157463.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423745.4\n",
      "selectiv5 \tselectiv3 \n",
      "1629.5\n",
      "selectiv5 \tselectiv3 \n",
      "1236952.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423814.6\n",
      "selectiv5 \tselectiv3 \n",
      "913848.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423722.3\n",
      "selectiv5 \tselectiv3 \n",
      "179838.8\n",
      "selectiv5 \tselectiv3 \n",
      "1423767.3\n",
      "selectiv5 \tselectiv3 \n",
      "1236545.3\n",
      "selectiv5 \tselectiv3 \n",
      "1349780.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423809.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423778.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423785.4\n",
      "selectiv5 \tselectiv3 \n",
      "396598.3\n",
      "selectiv5 \tselectiv3 \n",
      "1218173.3\n",
      "selectiv5 \tselectiv3 \n",
      "586220.3\n",
      "selectiv5 \tselectiv3 \n",
      "33966.9\n",
      "selectiv5 \tselectiv3 \n",
      "585520.4\n",
      "selectiv5 \tselectiv3 \n",
      "1437596.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423719.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423724.4\n",
      "selectiv5 \tselectiv3 \n",
      "557436.17\n",
      "selectiv5 \tselectiv3 \n",
      "1437610.3\n",
      "selectiv5 \tselectiv3 \n",
      "888027.8\n",
      "selectiv5 \tselectiv3 \n",
      "83684.12\n",
      "selectiv5 \tselectiv3 \n",
      "1423739.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423812.3\n",
      "selectiv5 \tselectiv3 \n",
      "1122228.3\n",
      "selectiv5 \tselectiv3 \n",
      "1131731.3\n",
      "selectiv5 \tselectiv3 \n",
      "872327.3\n",
      "selectiv5 \tselectiv3 \n",
      "1583.306\n",
      "selectiv5 \tselectiv3 \n",
      "1423720.3\n",
      "selectiv5 \tselectiv3 \n",
      "547043.6\n",
      "selectiv5 \tselectiv3 \n",
      "1585.83\n",
      "selectiv5 \tselectiv3 \n",
      "1423798.3\n",
      "selectiv5 \tselectiv3 \n",
      "1437609.3\n",
      "selectiv5 \tselectiv3 \n",
      "862514.3\n",
      "selectiv5 \tselectiv3 \n",
      "1505727.12\n",
      "selectiv5 \tselectiv3 \n",
      "525328.4\n",
      "selectiv5 \tselectiv3 \n",
      "203120.23\n",
      "selectiv5 \tselectiv3 \n",
      "542833.3\n",
      "selectiv5 \tselectiv3 \n",
      "1236950.3\n",
      "selectiv5 \tselectiv3 \n",
      "1255.195\n",
      "selectiv5 \tselectiv3 \n",
      "585524.3\n",
      "selectiv5 \tselectiv3 \n",
      "1341694.4\n",
      "selectiv5 \tselectiv3 \n",
      "525337.3\n",
      "selectiv5 \tselectiv3 \n",
      "54911.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423761.3\n",
      "selectiv5 \tselectiv3 \n",
      "337330.69\n",
      "selectiv5 \tselectiv3 \n",
      "557436.4\n",
      "selectiv5 \tselectiv3 \n",
      "1211775.3\n",
      "selectiv5 \tselectiv3 \n",
      "1122153.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423743.5\n",
      "selectiv5 \tselectiv3 \n",
      "1423785.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423816.3\n",
      "selectiv5 \tselectiv3 \n",
      "390333.7\n",
      "selectiv5 \tselectiv3 \n",
      "518634.7\n",
      "selectiv5 \tselectiv3 \n",
      "555970.3\n",
      "selectiv5 \tselectiv3 \n",
      "1122148.6\n",
      "selectiv5 \tselectiv3 \n",
      "390333.13\n",
      "selectiv5 \tselectiv3 \n",
      "1123687.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423829.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423747.3\n",
      "selectiv5 \tselectiv3 \n",
      "1435051.3\n",
      "selectiv5 \tselectiv3 \n",
      "585506.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423798.4\n",
      "selectiv5 \tselectiv3 \n",
      "203120.7\n",
      "selectiv5 \tselectiv3 \n",
      "525367.9\n",
      "selectiv5 \tselectiv3 \n",
      "1255.14\n",
      "selectiv5 \tselectiv3 \n",
      "1226298.9\n",
      "selectiv5 \tselectiv3 \n",
      "1616.3\n",
      "selectiv5 \tselectiv3 \n",
      "913848.7\n",
      "selectiv5 \tselectiv3 \n",
      "1423781.4\n",
      "selectiv5 \tselectiv3 \n",
      "1291764.5\n",
      "selectiv5 \tselectiv3 \n",
      "993692.3\n",
      "selectiv5 \tselectiv3 \n",
      "33962.8\n",
      "selectiv5 \tselectiv3 \n",
      "390333.12\n",
      "selectiv5 \tselectiv3 \n",
      "1423726.3\n",
      "selectiv5 \tselectiv3 \n",
      "1435464.3\n",
      "selectiv5 \tselectiv3 \n",
      "1450649.3\n",
      "selectiv5 \tselectiv3 \n",
      "1437595.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423833.6\n",
      "selectiv5 \tselectiv3 \n",
      "82380.43\n",
      "selectiv5 \tselectiv3 \n",
      "1423738.3\n",
      "selectiv5 \tselectiv3 \n",
      "1462575.3\n",
      "selectiv5 \tselectiv3 \n",
      "47770.84\n",
      "selectiv5 \tselectiv3 \n",
      "1437611.3\n",
      "selectiv5 \tselectiv3 \n",
      "1494462.5\n",
      "selectiv5 \tselectiv3 \n",
      "1341695.4\n",
      "selectiv5 \tselectiv3 \n",
      "557433.4\n",
      "selectiv5 \tselectiv3 \n",
      "1122149.3\n",
      "selectiv5 \tselectiv3 \n",
      "47770.576\n",
      "selectiv5 \tselectiv3 \n",
      "1423786.4\n",
      "selectiv5 \tselectiv3 \n",
      "46256.17\n",
      "selectiv5 \tselectiv3 \n",
      "1236546.3\n",
      "selectiv5 \tselectiv3 \n",
      "1410645.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423799.3\n",
      "selectiv5 \tselectiv3 \n",
      "525330.7\n",
      "selectiv5 \tselectiv3 \n",
      "518634.21\n",
      "selectiv5 \tselectiv3 \n",
      "1423784.4\n",
      "selectiv5 \tselectiv3 \n",
      "1410647.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423833.5\n",
      "selectiv5 \tselectiv3 \n",
      "391904.8\n",
      "selectiv5 \tselectiv3 \n",
      "1423744.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423782.3\n",
      "selectiv5 \tselectiv3 \n",
      "1629.13\n",
      "selectiv5 \tselectiv3 \n",
      "480391.6\n",
      "selectiv5 \tselectiv3 \n",
      "1578.21\n",
      "selectiv5 \tselectiv3 \n",
      "1423764.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423721.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423815.3\n",
      "selectiv5 \tselectiv3 \n",
      "1437608.4\n",
      "selectiv5 \tselectiv3 \n",
      "1293592.5\n",
      "selectiv5 \tselectiv3 \n",
      "566552.17\n",
      "selectiv5 \tselectiv3 \n",
      "1423746.3\n",
      "selectiv5 \tselectiv3 \n",
      "703613.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423762.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423727.3\n",
      "selectiv5 \tselectiv3 \n",
      "137591.149\n",
      "selectiv5 \tselectiv3 \n",
      "1423780.4\n",
      "selectiv5 \tselectiv3 \n",
      "566552.16\n",
      "selectiv5 \tselectiv3 \n",
      "1624.128\n",
      "selectiv5 \tselectiv3 \n",
      "1423764.5\n",
      "selectiv5 \tselectiv3 \n",
      "518634.19\n",
      "selectiv5 \tselectiv3 \n",
      "525330.3\n",
      "selectiv5 \tselectiv3 \n",
      "1255.124\n",
      "selectiv5 \tselectiv3 \n",
      "1045004.5\n",
      "selectiv5 \tselectiv3 \n",
      "218139.4\n",
      "selectiv5 \tselectiv3 \n",
      "109790.239\n",
      "selectiv5 \tselectiv3 \n",
      "1423742.4\n",
      "selectiv5 \tselectiv3 \n",
      "1423833.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423718.3\n",
      "selectiv5 \tselectiv3 \n",
      "1121107.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423760.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423782.4\n",
      "selectiv5 \tselectiv3 \n",
      "1220560.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423766.4\n",
      "selectiv5 \tselectiv3 \n",
      "518635.7\n",
      "selectiv5 \tselectiv3 \n",
      "372484.4\n",
      "selectiv5 \tselectiv3 \n",
      "1410647.3\n",
      "selectiv5 \tselectiv3 \n",
      "324831.13\n",
      "selectiv5 \tselectiv3 \n",
      "1423779.3\n",
      "selectiv5 \tselectiv3 \n",
      "1423817.3\n",
      "selectiv5 \tselectiv3 \n",
      "1680.108\n",
      "selectiv5 \tselectiv3 \n"
     ]
    }
   ],
   "source": [
    "def prepare_aflp_silco():\n",
    "    os.chdir('AFLPinSilico')\n",
    "    folder = '../../Data/BacDive/AFLP_perl/'\n",
    "\n",
    "    if not os.path.exists(folder):\n",
    "        os.makedirs(folder)\n",
    "\n",
    "    # !cpan CGI -y # In case the CGI module is not installed\n",
    "\n",
    "    if sys.platform.startswith('win'):\n",
    "        perl_script = 'script_AFLP_perl.pl'\n",
    "    elif sys.platform.startswith('darwin') or sys.platform.startswith('linux'):\n",
    "        perl_script = 'script_AFLP_perl_UNIX.pl'\n",
    "    else:\n",
    "        print(\"Operative System not supported.\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    os.system(f\"perl {perl_script}\")\n",
    "\n",
    "    files_name = sorted([os.path.basename(f).replace('.wri', '') for f in glob.glob(folder + '*.wri')])\n",
    "    \n",
    "    os.chdir('..')\n",
    "    \n",
    "    return files_name\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def generate_dataset(files_name):\n",
    "    folder = '../Data/BacDive/AFLP_perl/'\n",
    "    \n",
    "    aflp_int = list()\n",
    "    file_col = list()\n",
    "\n",
    "    for file in files_name:\n",
    "\n",
    "        file_col.append(file[:13])\n",
    "        aflp_int_row = list()\n",
    "\n",
    "        aflp_50_500 = pandas.read_table(folder+file+'.wri', header=10)\n",
    "        aflp_50_500.rename(columns={'Unnamed: 0': 'ID'}, inplace=True)\n",
    "        aflp_50_500 = aflp_50_500.loc[(aflp_50_500['FragLength'] >= 50) & (aflp_50_500['FragLength'] <= 500)]\n",
    "\n",
    "        freq = aflp_50_500['FragLength'].value_counts()\n",
    "        frag_length = freq.sort_index()\n",
    "        frag_length_index = frag_length.index\n",
    "\n",
    "        for f in range(50,501,1):\n",
    "            if f in frag_length_index:\n",
    "                aflp_int_row.append(freq[f])\n",
    "            else:\n",
    "                aflp_int_row.append(0)\n",
    "\n",
    "        aflp_int.append(aflp_int_row)\n",
    "\n",
    "    aflp_int_df = pandas.DataFrame(aflp_int)\n",
    "    aflp_int_df.columns = list(range(50, 501))\n",
    "    aflp_int_df['sequence_genomes.Seq. accession number'] = file_col\n",
    "    aflp_int_df.to_excel('../Data/BacDive/aflp_int_AFLPinSilco.xlsx',index=False)\n",
    "    \n",
    "    # Reads the original dataset to eliminate superfluous information, set the indexes and to do some checks.\n",
    "    X = pandas.read_excel(\"../Data/BacDive/aflp_int_AFLPinSilco.xlsx\")\n",
    "    y = pandas.read_excel(\"../Data/BacDive/Genome_50CH_ac_0_1.xlsx\")\n",
    "\n",
    "    X['sequence_genomes.Seq. accession number'] = X['sequence_genomes.Seq. accession number'].astype(str)\n",
    "    y['sequence_genomes.Seq. accession number'] = target['sequence_genomes.Seq. accession number'].astype(str) \n",
    "\n",
    "    X = X.rename(columns={'sequence_genomes.Seq. accession number': 'strain'}).set_index('strain').sort_index()\n",
    "    y = y.rename(columns={'sequence_genomes.Seq. accession number': 'strain'}).set_index('strain').sort_index()\n",
    "\n",
    "    y = y.iloc[:, list(range(11, 61))]\n",
    "    y = y.loc[X.index]\n",
    "\n",
    "    for i in range(len(X.index)):\n",
    "        if X.index[i] != y.index[i]:\n",
    "            print(f\"The row indices in row {i} are not equal: {X.index[i]} vs {y.index[i]}\")\n",
    "\n",
    "    with pandas.ExcelWriter('../Data/X.xlsx') as writer:\n",
    "        X.to_excel(writer, index=True)\n",
    "\n",
    "    with pandas.ExcelWriter('../Data/Y.xlsx') as writer:\n",
    "        y.to_excel(writer, index=True)\n",
    "        \n",
    "    print(\"\\n\\nDataset saved in Data/X.xlsx and in Data/Y.xlsx\")\n",
    "    \n",
    "    display(X)\n",
    "    display(y)\n",
    "\n",
    "    \n",
    "    \n",
    "files_name = prepare_aflp_silco()\n",
    "generate_dataset(files_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
